Add gene set heatmap tracks to the Oncoprint

fedde-s opened the pull request:
If the user queries gene sets and a gene set profile in addition to
genes, add the corresponding tracks to the bottom of the Oncoprint with
an appropriate rule set.

## Screenshot ##
![Screenshot of an Oncoprint showing some genetic tracks, some expression heatmap tracks, and below that, two gene set tracks in a green-to-dark-pink scale](https://user-images.githubusercontent.com/4929431/35055646-efb79df8-fbaf-11e7-9fdf-21151e61a7b9.png)

fedde-s commented:
This feature ports the very basics of the functionality gathered in the now-defunct pull request cBioPortal/cbioportal#3319. Using it depends on a backend with the tweaks proposed in pull request cBioPortal/cbioportal#3735, and on studies containing gene set data (such as the example in cBioPortal/datahub#45).

adamabeshouse commented on src/pages/resultsView/ResultsViewPageStore.ts:
```diff
@@ -1323,6 +1328,26 @@ export class ResultsViewPageStore {
+            this.molecularProfilesInStudies
+        ],
+        invoke: () => {
+            const GENESET_SCORE = "GENESET_SCORE";
```
add this to `AlterationTypeConstants` in this file

fedde-s replied:
Ah, I see, that was added last week, before I wrote this code. Yes, that does sound like a sensible place for it 👍

adamabeshouse commented on src/shared/components/oncoprint/DataUtils.spec.ts:
```diff
@@ -1,12 +1,16 @@
 import {MutationSpectrum} from "../../api/generated/CBioPortalAPIInternal";
 
+/* Type assertions are used throughout this file to force functions to accept
+/* mocked parameters known to be sufficient. */
```
:+1:

adamabeshouse commented on src/shared/components/oncoprint/OncoprintUtils.ts:
```diff
@@ -31,15 +42,48 @@ export function doWithRenderingSuppressedAndSortingOff(oncoprint:OncoprintJS<any
+         * The PiYG colormap is based on color specifications and designs
+         * developed by Cynthia Brewer (http://colorbrewer.org).
+         * The palette has been included under the terms
+         * of an Apache-style license.
```
I actually barely even know what you mean by that, let alone know the answer :D paging @alisman

fedde-s commented:
The pre-commit hooks seem to have changed since the last time I rebased on `rc`, and the hook named ‘test’ is now failing. The only explanation I can see is a long list of filenames followed by a coverage summary, so I assumed I should ignore the hook.

fedde-s commented:
Since [Adam's review](https://github.com/cBioPortal/cbioportal-frontend/pull/898#pullrequestreview-89979757) I added a test to cover how `fillHeatmapTrackDatum()` behaves differently for gene sets (which isn't too much, and mostly also covered by the type system), moved a constant to a central place [as requested](https://github.com/cBioPortal/cbioportal-frontend/pull/898#discussion_r162508681), and, for now, bundled the license in a file similar to the one I'd used in the old project. I think that covers everything that has come up in review so far.

adamabeshouse commented:
@fedde-s it looks great! I think some end-to-end (screenshot) tests would be great too. Sorry to put this through the wringer, but I just want to be rigorous because our testing practices have become more rigorous recently, and this is a completely new feature.

Unfortunately, our end-to-end testing on Browserstack is having some issues w webGL and so oncoprint tests aren't working. We submitted a ticket about it to Browserstack and we're waiting on a response. Let me get back to you with those details hopefully within the next couple of days, and we can finish this up and merge it.

fedde-s commented:
I asked some people here to point me to a study contains suitable GSVA data for such a test. That will have to be loaded in the back-end used by the test; not sure where that is. I will also report back when I know which study to load! 😄

alisman commented:
@adamabeshouse what version of oncoprintjs is required for this?   can we update package.json/lock in this PR?

adamabeshouse commented:
@fedde-s I merged your changes to the oncoprint repo 🎉  can you please update the package.json oncoprintjs version to 1.0.79? (`npm install --save oncoprintjs@1.0.79`)

alisman commented on src/pages/resultsView/ResultsViewPageStore.ts:
```diff
@@ -1313,6 +1319,25 @@ export class ResultsViewPageStore {
+            this.molecularProfilesInStudies
+        ],
+        invoke: () => {
+            const applicableProfiles = _.filter(
```
We should move this to a helper function and test it

fedde-s replied:
What would be a sensible place for such a helper function? The top level of the same module?

fedde-s replied:
I think there's two other things wrong with this function. First, I think the array I intended to use is `selectedMolecularProfiles` instead of `molecularProfilesInStudies`, as I meant to select only queried profiles. And second, it can be a bit confusing and difficult to explicitly typecheck a MobxPromise that resolves with `undefined` since its `.result` property is also undefined if the promise hasn't resolved (and usually presumed not to be after it has). I'll add two commits to address this.

fedde-s commented:
@adamabeshouse and @alisman, the functionality in this PR does not depend on any of the features in cBioPortal/oncoprintjs#34; I wrote and ran it using the OncoprintJS version already present in rc. The functionality that will depend on the features is expanding the gene sets to reveal the genes, which will be another pull request.

sandertan commented:
Test data can be found in https://github.com/cBioPortal/datahub/pull/156. We calculated GSVA scores and p-values for the `gbm_tcga` study. @inodb @alisman could one of you load this study in the appropriate test database? A link to documentation for loading gene set data can be found in the description of that PR.

fedde-s commented:
As the underlying `rc` branch had changed quite a bit, I rebased my changes on its current state.

fedde-s commented:
I pushed two commits to fix mistakes I made around the logic to select the queried gene set profile: I used all profiles in the studies as input instead of starting with only the queried profiles as I intended, and I used `undefined` to mean two different things, confusing myself into thinking it would not be undefined further down the line.

fedde-s commented:
If you're okay with the four amendments I could squash them into the original commit.

fedde-s commented:
Two more features are waiting for this to be merged so far: expansion of the gene tracks (see thehyve/cbioportal-frontend#2, I'll make a PR to rc once this is merged) and links to the gene set metadata (see thehyve/cbioportal-frontend#3). I'm working on a third one that will enable clustering for gene set heatmaps. All this functionality has previously been reviewed in the pull requests listed in cBioPortal/cbioportal#3319.

fedde-s commented:
Pushed a commit to render the gene set heatmap in the track group immediately following the last one used by a dynamically queried heatmap (or the genetic tracks) at any point, to address [Adam's ‘hmm’](https://github.com/cBioPortal/cbioportal-frontend/pull/898#discussion_r167972888) and further feedback over other media.

fedde-s commented:
The ported logic for the third feature depending on this one can be found here: thehyve/cbioportal-frontend#4. Again, I'll make pull requests to `rc` once this one is in there.

Sjoerd-van-Hagen commented:
Hi @adamabeshouse @alisman,

We would like to get this pull request and the related PR's merged as soon as possible. Could you please comment on anything that really needs to be changed in this pull request before it can be merged?

Next we would like to implement the related merged tracks feature and we are likely able to add some extra tests and make further improvements to the code there.

Thanks!

Sjoerd.

fedde-s commented:
Rebased on `rc` to resolve conflicts.

adamabeshouse commented:
Hi @Sjoerd-van-Hagen we just were able to get the test data up and running in a DB, we'll test the feature this afternoon and let you know of any final changes if any come up, I don't expect anything else will be necessary if we dont find any bugs.

Sjoerd-van-Hagen commented:
Hi @adamabeshouse,

Great! Thanks, and good luck testing!

adamabeshouse commented:
@fedde-s Looks good! A couple of things I'm noticing - for all of the below, I'm talking about running cbioportal/rc and using frontend for this branch (rebased onto rc), connecting to the database you sent.

(1) This makes it possible to submit a query without anything in the main OQL box, but only in the gene set box. 

A lot of our code makes the implicit assumption that at least one actual gene is selected, so this causes bugs. You either need to make it mandatory to select some genes in the OQL box, or go through the code and fix the bugs that come up when submitting a query with no gene.

(2) Some queries fail with an error like - no permission to INSERT into `text_cache`. I can't seem to reproduce it but it definitely happened once and refreshing the page didn't help.

(3) The genesets box does not populate in the modify query form eg in `index.do?cancer_study_id=gbm_tcga&Z_SCORE_THRESHOLD=2.0&RPPA_SCORE_THRESHOLD=2.0&data_priority=0&case_set_id=gbm_tcga_all&gene_list=PTEN&geneset_list=VERHAAK_GLIOBLASTOMA_CLASSICAL&tab_index=tab_visualize&Action=Submit&genetic_profile_ids_PROFILE_MUTATION_EXTENDED=gbm_tcga_mutations&genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=gbm_tcga_gistic&genetic_profile_ids_PROFILE_GENESET_SCORE=gbm_tcga_gsva_scores`

fedde-s commented:
Ah, the thing where the query page allows queries with no genes was added in #742, expecting that we'd be able to adjust the code to make that possible just like we originally did in the old frontend. That's probably a bit too much work to do right now though, and it would be better to disallow such queries until we do have time for that. @oplantalech, would it be easy for you to restore that one error case?

fedde-s commented:
Rebased again to adjust to changes in `rc`.

fedde-s commented:
The empty modify query box is another already present in `rc` via #742, we should add an issue for that.

fedde-s commented:
As for the `text_cache` warning, I was not aware that the web server would ever need to write to the database either, but it seems like it was added by @onursumer many years ago. Maybe he could clarify whether the error message is a problem; I've never seen it.

Sjoerd-van-Hagen commented:
Hi @adamabeshouse,

Thanks for reviewing and merging this PR!

Cheers,

Sjoerd.
